File: beg++17/ch02/enums.cpp
// enums.cpp

// A complete C++ Program
#include <iostream>

enum class Color: unsigned int {
     Red = 0xFF0000,
     Green = 0x008000,
     Yellow = 0xFFFF00,
     Purple = 0x800080,
     Blue = 0x0000FF,
     Black = 0x000000,
     White = 0xFFFFFF
};

void print_rgb(Color);

int main(int argc, char** argv)
{
    Color colors[] = {Color::Yellow, Color::Purple, Color::Green};
    for (Color color: colors) {
        print_rgb(color);
    }

    return 0;
}

void print_rgb(Color color) {
   unsigned int value = static_cast<unsigned int>(color);

    unsigned int red = (value >> 16) & 0xFF;
    unsigned int green = (value >> 8) & 0xFF;
    unsigned int blue = value & 0xFF;

    std::cout << "RGB(" << red << ", " << green << ", "
              << blue<<")" << std::endl;
}

File: beg++17/ch02/enums.rs
#![allow(dead_code, unused)]

#[derive(Debug)]
enum Color {
    Red = 0xFF0000,
    Green = 0x008000,
    Yellow = 0xFFFF00,
    Purple = 0x800080,
    Blue = 0x0000FF,
    Black = 0x000000,
    White = 0xFFFFFF,
}

fn main() {
    let red = Color::Red;
    print_rgb(red);
    let green = Color::Green;
    print_rgb(green);
}

fn print_rgb(color: Color) {
    let value: u32 = color as u32;

    let red = (value >> 16) & 0xFF;
    let green = (value >> 8) & 0xFF;
    let blue = value & 0xFF;

    println!("RGB({}, {}, {})", red, green, blue);
}
